### PR TITLE
AVSR recipes on LRS3 using pre-trained AV-HuBERT model

### PR DESCRIPTION
**Pull request by ms-dot-k**
## What?

New recipe for training audio-visual speech recognition model using LRS3 dataset added egs2/lrs3/avsr1 .
Recipe is using dumped features using pre-trained frontends of AV-HuBERT.
Therefore, the dumped features are the fusion of audio-visual features.

## Why?

Create training recipe for AVSR using AV-HuBERT model.

## See also

As we utilize dumped feature, we cannot utilize some data augmentations original used in AV-HuBERT.
For example, audio noise modeling & video random cropping and flipping.
Instead, the recipe applies random time masking at audio-visual features. Moreover, to avoid overfitting, it utilizes more light transformer decoder when AV-HuBERT Large configuration is used. 
(Original AV-HuBERT Large uses 8 heads with 9 layers while the recipe uses 4 heads with 6 layers for decoder.)

### CODE REVIEW

File: espnet2/asr/encoder/avhubert_encoder.py [hunk 1]
```diff
@@ -0,0 +1,1092 @@
+# Copyright 2023
```
Comment (ftshijt) [inline espnet2/asr/encoder/avhubert_encoder.py:1]: Many thanks for making it working by adapting the original codebase of fairseq to here. My major concern is that the fairseq-style is not very compatible to our style (introducing a lot of module-related codes, which makes the design a bit redundant)

I feel that there would be two better solutions:
- Option1: follow https://github.com/ms-dot-k/espnet/blob/master/espnet2/asr/encoder/wav2vec2_encoder.py style that directly import fairseq?
- Option2: add an entry to s3prl (which we usually use for SSL and they have a good capacity of holding different SSL models) https://github.com/s3prl/s3prl

Comment (ftshijt) [reply espnet2/asr/encoder/avhubert_encoder.py:1]: I'm open for discussion. Maybe also consult @simpleoier and @sw005320 's thoughts on it

**Review by simpleoier — COMMENTED** [review]
Thanks for your efforts!
I left some comments.

File: egs2/lrs3/avsr1/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+- Git hash: `5d0758e2a7063b82d1f10a8ac2de98eb6cf8a352`
+  - Commit date: `Wed Aug 30 18:03:42 2023 -0400`
+- Pretrained model: [https://huggingface.co/espnet/msk_lrs3_train_avsr_avhubert_large_extracted_en_bpe1000](https://huggingface.co/espnet/msk_lrs3_train_avsr_avhubert_large_extracted_en_bpe1000)
+
+## exp/asr_train_avsr_avhubert_large_extracted_en_bpe1000
+### WER
+
+|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
+|---|---|---|---|---|---|---|---|---|
+|inference_asr_model_valid.acc.ave/test|1321|9890|98.5|1.1|0.4|0.2|1.7|8.8|
```
Comment (sw005320) [inline egs2/lrs3/avsr1/README.md:17]: It is very good.
Is it an expected number?
The performance of this task is already saturated?

File: egs2/lrs3/avsr1/local/extract_av_feature.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+
+def stacker(feats, stack_order):
+    """
+    Concatenating consecutive audio frames
+    Args:
+    feats - numpy.ndarray of shape [T, F]
+    stack_order - int (number of neighboring frames to concatenate
+    Returns:
+    feats - numpy.ndarray of shape [T', F']
```
Comment (sw005320) [inline egs2/lrs3/avsr1/local/extract_av_feature.py:57]: Would F' be changed from F?
If not, maybe [T', F]?

File: espnet2/asr/encoder/avhubert_encoder.py (hunk 1 above)
Comment (sw005320) [reply espnet2/asr/encoder/avhubert_encoder.py:1]: I think it's better to copy it instead of importing it from fairseq.
We had a lot of issues when importing fairseq.
Also, this approach allows us to edit the model directly (e.g., adding adapter, etc.)

File: espnet2/asr/encoder/avhubert_encoder.py
Comment (sw005320) [inline espnet2/asr/encoder/avhubert_encoder.py:1]: We should add a test to cover this part.
Please get an instruction from @ftshijt or @simpleoier about it.

Comment (sw005320) [reply espnet2/asr/encoder/avhubert_encoder.py:1]: Hmm, I see that you added a test.
I don't know why but from my view, it looks like this code does not go through the test.
I'll check it later again.

**Review by sw005320 — COMMENTED** [review]
LGTM.

Maybe we could add a bit more comments (e.g., how to extract video features, how to fuse them, how to feed them to espnet (through the Kaldi ark format?), and how to use av hubert) to this recipe in `egs2/lrs3/avsr1/README.md.`

File: egs2/lrs3/avsr1/local/data.sh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+log() {
+    local fname=${BASH_SOURCE[1]##*/}
+    echo -e "$(date '+%Y-%m-%dT%H:%M:%S') (${fname}:${BASH_LINENO[0]}:${FUNCNAME[1]}) $*"
+}
+
+. ./db.sh
+. ./path.sh
+. ./cmd.sh
+
+cmd=run.pl
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply egs2/lrs3/avsr1/local/data.sh:17]: Yeah, in some scripts, `run.pl` is used as the default value of `cmd`. But in that case, the cmd is passed as an argument from another script. Also, `./cmd.sh` above is not necessary.

File: egs2/lrs3/avsr1/local/extract_av_feature.py (hunk 3 above)
Comment (sw005320) [reply egs2/lrs3/avsr1/local/extract_av_feature.py:57]: Oh, got it!
maybe, you can add these descriptions here?

**Comment by sw005320** [comment]
Thanks a lot for your great efforts!
This was our long-term action item to add an audio-visual ASR task!

The last comment is a follow-up PR to add this great new function to https://github.com/espnet/espnet/blob/master/README.md
Please add some documents and advertise this great work!
